sycl/plugins/unified_runtime/ur/adapters/level_zero/usm.cpp

```diff
@@ -378,58 +428,95 @@ static ur_result_t USMFreeImpl(ur_context_handle_t Context, void *Ptr) {
   return UR_RESULT_SUCCESS;
 }
 
-void *USMMemoryAllocBase::allocate(size_t Size) {
-  void *Ptr = nullptr;
+static ur_result_t USMQueryPageSize(ur_context_handle_t Context,
```

Comment:
@jandres742 Do you know if this is a valid way to query L0 for page size?

Reply:
thanks @kswiecicki . This is a good pint. L0 doesnt have a good way to query for the page size. 

if pointer is already allocated, then it is easy (that's the if(ptr) branch below).

however, if pointer is null, L0 doesnt have an API to return the page size (this has been discussed in the past and such API has been rejected, given that the actual page depends on final allocation, so L0 cannot give guarantees on page size before alloction).

The second branch below that uses zeVirtualMemQueryPageSize is designed to be used for knowing the pagesize to be used for alignment of requested virtual address range. However, that is only for device allocations.

Which begs the question, what is the objective of:

```cpp
umf_result_t USMMemoryProvider::get_min_page_size(void *Ptr, size_t *PageSize) {
```

which is the one calling USMQueryPageSize? 

if it is to query the pagesize of a non-null ptr, then no issues. But if it is to query for the minimum page size supported, how adapter knows whether user is referring to host, device, or shared, in case those have different page size requirements? how this is handled in CUDA?

Reply:
We wanted to use this mechanism in the implementation of a heap manager. Knowing what the min page size is would make it easier to handle alignment (i.e. we don't need to pass any specific alignment to zeMemAllocShared, which could potentially waste some space).

Regarding whether user is referring to host, device, or shared memory: that function would be implemented for each memory provider (i.e for USMShared, USMDevice and USMHost providers) separately. 

However, If L0 does not support this (and I believe CUDA does not as well) we should probably require the `Ptr` to be not null for this function. This will require some changes in the heap manager as well but nothing major.

Reply:
thanks @igchor . another way of doing this for null ptrs, would be to create a small allocation (zeMemAllocHost/Device/ or Shared), call zeMemGetAllocProperties, and get the pagesize. This is what I think we are doing in other libraries. It has the disadvantage of needing to allocate and free the allocation when calling USMQueryPageSize, but it has the advantage of providing an accurate result. So:

```cpp
USMQueryPageSize(ptr) {

  void *usmPtr = ptr;
  if (ptr == nullptr) {
    zeMemAlloc*(&usmPtr);
  }
 
  zeMemGetAllocProperties(usmPtr);
  // get pagesize

  if (ptr == nullptr) {
    zeMemFree(&usmPtr);
  }

}
```

Reply:
That's a good point, thank you - we could use something like this in the heap manager implementation. 

I'm wondering if we should do something like this inside the `get_min_page_size` automatically or it would be better to require the `ptr` to be != NULL and let the user explicitly do the allocation, query, and deallocation. We don't put any requirements on `get_min_page_size` currently and I think this flow should work for other adapters as well, so we probably could do that automatically...

Unless you think that having this allocation could be problematic in some cases?

Reply:
@igchor the problem would be if the user calls this function a lot of times, which is why caching the pagesize is important. I think that not requiring an allocation from the user is better, which is the natural thinking ("I want to allocate in the future, what is the minimal page size?"). The hidden allocation is just because this cannot be provided at the moment (but maybe L0 could do it in the future, or other adapters can provide it).

@kbenzie : what do you think?

Reply:
Sorry I missed this before now.

In generally if we can avoid hidden allocations that would ideal but if this is necessary to get the min page size at all then caching the min page size to avoid overhead of multiple hidden allocations is a reasonable middle ground.

What sorts of ranges of pages sizes are we talking?

Reply:
@kbenzie : we only need an allocation of 1byte to get minimal page size.